### PR TITLE
YJIT: Use starting context for status === CantCompile

### PR DESCRIPTION
**Pull request by jimmyhmiller**

### CODE REVIEW

**Comment by jimmyhmiller** [comment]
I looked into the failure. It is unrelated to my change. I can actually recreate it locally on master. I will make an issue about it.

**Comment by maximecb** [comment]
Thanks for fixing this and adding new tests.

File: bootstraptest/test_yjit.rb [hunk 1]
```diff
@@ -3690,3 +3690,45 @@ def literal_append(sql, v)
...
+    when 1
+      1
+    when 2
+      2
+    else
+      3
+    end
+  end
+
+  my_func(1)
```
Comment (k0kubun) [inline bootstraptest/test_yjit.rb:3713]: Do you remember how this test failed when you didn't use `starting_ctx`? I assume this is about `CantCompile` on `opt_case_dispatch` due to `Integer#===` redefinition, but it doesn't seem to push/pop stack operands before that `CantCompile`.

**Comment by k0kubun** [comment]
When we do stack temp register allocation, just looking at a stack operand (even without push/pop) can result in a register spill, i.e. a destructive ctx change. So I'm planning to remove all `starting_ctx` patterns, including this PR's change.

I'll work on it myself tomorrow, but if you already have insight as to which place should be fixed (is it just the `ctx.stack_pop(1)` for block arg?), your input would be appreciated.

File: bootstraptest/test_yjit.rb (hunk 1 above)
Comment (k0kubun) [reply bootstraptest/test_yjit.rb:3713]: well, sorry, never mind. I was on a branch that removed `starting_ctx` from `opt_case_dispatch` 🙂 That's why.

**Comment by maximecb** [comment]
> Another idea: If we store asm.insns.len() before compiling each YARV insn, when CantCompile is returned, maybe we could rewind asm to the point before the code for the cant-compile YARV insn is generated. Then we could safely use starting_ctx even with stack temp register allocation.

I'm a bit confused.

Normally, we return `CantCompile` before any instructions are generated? Is this not the case here?

**Comment by k0kubun** [comment]
> Normally, we return CantCompile before any instructions are generated?

This doesn't seem like the case for method calls. `invokesuper` and `invokeblock` generating guards before calling `gen_send_iseq` (which may return `CantCompile`) are fine-ish because it doesn't touch the stack. However, for `OPTIMIZED_METHOD_TYPE_SEND` (`send(:foo, bar)`), you inevitably need to look at the stack first and then go through the normal method call sequence (which may return `CantCompile`) again. It doesn't seem trivial to implement the assumption without rewinding an Assembler.

**Comment by k0kubun** [comment]
I can still look at simply not using `starting_ctx` for `CantCompile` though. It's probably too late to avoid generating code before `CantCompile`, but avoiding `stack_pop` before `CantCompile` might be easier than that.

**Pull request merged**